pipelines: test: ldd-check: cleanups

dannf opened the pull request:
Ref: https://github.com/wolfi-dev/os/pull/38563/files#r1906010589

dannf committed:
pipelines: test: ldd-check: factory out test for dynamic objects

Error out if ldd was not able to actually reason about whether or not
a file is a dynamic object by checking it's error message.

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>

smoser commented on pipelines/test/ldd-check.yaml:
```diff
@@ -62,17 +62,22 @@ pipeline:
+      is_dyn_obj() {
+        # ldd has this code already, let's not NIH it
+        # note: the actual ldd check happens in test_file()
+        ldd "$1" > /dev/null 2>&1
```
Does it by chance edit with a different code for "not a dynamic object" than for "catastrophic failures"

dannf replied:
It does not. But we can check it's error message. See the updated commit.

smoser commented on pipelines/test/ldd-check.yaml:
```diff
@@ -62,6 +62,19 @@ pipeline:
         fail "$f: missing ${missing# }"
```
huh. i think the above line here should be 'return 0', not 'continue' as we're not in a loop here.

smoser committed:
adjust ldd-check

I started down this route just trying to fix 'continue'
being called from a function (ratner than return).

Then the nerd-snype / odd love for pain in shell hit me.

Other things here:
1. just a single call to 'ldd' for each file.
2. if 'insist_dyn' is false, then we will just log a message
   instead of failing in the following cases:
   * not a file
   * does not exist
   * is not a dynamic executable

   This allows you to call it on all contents in a package
   and mostly get the results  you want.
3. error if ldd exits other than 1 or 0 , and dump the output.
4. error (which exits immediately) if a package in 'packages'
   is not installed.
5. Please can we add a way to write pipelines in something else?

smoser commented:
I pushed some changes. please @dannf and others review.

murraybd commented on pipelines/test/ldd-check.yaml:
```diff
@@ -38,51 +38,94 @@ pipeline:
         test_files_in "$pkg"
       done
-      echo "tested $((passes+fails)) files with ldd. $passes passes. $fails fails."
+      echo "tested $((passes+fails)) files with ldd." \
```
I really wish this  line was prefaced with "[ldd-check]" like some of the other lines so one could easily grep log files for "[ldd-check]".

dannf replied:
https://github.com/wolfi-dev/os/pull/39042/commits/69f73a83c353e185990e438850681eeaffa657f3

dannf committed:
pipelines: test/ldd-check: include 'ldd-check' in summary message

Signed-off-by: dann frazier <dann.frazier@chainguard.dev>
